### ISSUE TITLE
Combine Variable and Tensor APIs (Perform autograd directly on torch.Tensor)

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#2633]
There are previous issues addressing this, but I want to write this up with a plan in a new issue.

We should make `torch.Tensor` differentiable instead of having a wrapper `Variable` class. The Variable class adds conceptual overhead without sufficient benefit. It makes it harder for new users to adopt PyTorch by adding a second API which is similar but not exactly the same as Tensor.

There are few prerequisites:

1. Operations on `Variable`s need to be at least as fast as operations on `Tensor` (https://github.com/pytorch/pytorch/issues/2630)
2. We need scalar support at the Python level (https://github.com/pytorch/pytorch/issues/1433)

The `Variable` constructor will be deprecated and become a no-op.
The `Variable.data` accessor will be deprecated and return a view that has requires_grad=False.
The `torch.Tensor` constructors will take an optional keyword argument `requires_grad`

There will be a number of internal changes, but we shouldn't break most existing code:

1. For performance reasons, we will no longer build up the graph of operations by default when `requires_grad=False` unless the operation uses stochastic variables. We'll use a context manager or global variable to enable graph creation.
2. The new differentiable Tensor will need support all the operations currently on Tensor, including in-place and `out=` operations.
3. In-place operations on leaf Variables which require_grad are not currently supported. In the new API, the tensor's value will be updated and it will remain a leaf node. (This is important for optimizers, which update leaf Variables in-place).

There are probably a few other subtle differences between Tensor and Variable APIs. I'll update this issue as we address them.

See also:
https://github.com/pytorch/pytorch/issues/1384
https://github.com/pytorch/pytorch/issues/274

## Failing Tests

- test/test_autograd.py 23/738 # Greg will be fixing the tests, Tensor vs. Variable requires_grad
- test/test_cuda.py 2/1837  # Just test_broadcast from TestTorch
- test/test_dataloader.py OK
- test/test_distributed.py OK # error at top level
- test/test_distributions.py OK
- test/test_indexing.py OK
- test/test_jit.py OK
- test/test_legacy_nn.py OK
- test/test_multiprocessing.py OK
- test/test_nccl.py OK
- test/test_nn.py OK
- test/test_optim.py OK
- test/test_sparse.py 10/120 # missing sspaddmm, dsmm/mm, addmm
- test/test_torch.py 1/257 TestTorch.test_broadcast
- test/test_utils.py 2OK

**Comment by apaszke** [comment]
Properly handling in-place might not be as easy as it seems. Sometimes you actually need to move the Variable in the graph. Consider this case:
```python
x = Variable(torch.zeros(10, 10))
y = Variable(torch.randn(10), requires_grad=True)
x[2] = y
x.sum().backward()
assert y.grad.data.eq(torch.ones(10))
```

**Comment by ngimel** [comment]
One of the cases of differences between variable and tensor is computation of var/std: https://github.com/pytorch/pytorch/issues/2470#issuecomment-323206469

**Comment by vadimkantorov** [comment]
If this is put further, would there be ways for implementing lazy graph evaluation?

**Comment by Jiaming-Liu** [comment]
IMO Variable gives user a clear difference between graph node and data.
.data attribute and Variable constructor provide convenient and explicit ways to take control of the graph (or graphs).

**Comment by colesbury** [comment]
This is done. (Still some clean-up to do)

**Comment by carlosgmartin** [comment]
Is the conda PyTorch out of date? I'm getting the following error: `RuntimeError: torch.FloatTensor constructor doesn't accept any keyword arguments`.

**Comment by ezyang** [comment]
Yes, we haven't released 0.4, which has the Variable Tensor merge, yet
